### PR TITLE
ENH: Allow `np.r_` to accept 0d arrays

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes gh-9233

This only changes errors to non-errors, so is backwards compatible.

Would add a release note, but that would conflict with #10186

### CODE REVIEW

**Comment by eric-wieser** [comment]
Fixes #4559 too

File: numpy/lib/index_tricks.py [hunk 1]
```diff
@@ -312,21 +312,16 @@ def __getitem__(self, key):
...
-                        k1 = trans1d
-                        axes = defaxes[:k1] + defaxes[k2:] + \
-                               defaxes[k1:k2]
-                        newobj = newobj.transpose(axes)
-                    del tempobj
+                item_ndim = ndim(item)
+                newobj = array(item, copy=False, subok=True, ndmin=ndmin)
+                if trans1d != -1 and item_ndim < ndmin:
+                    k2 = ndmin - item_ndim
+                    k1 = trans1d
```
Comment (charris) [inline numpy/lib/index_tricks.py]: So the big difference is that `trans1d` is not updated? That seems like a big change.

Comment (eric-wieser) [reply numpy/lib/index_tricks.py]: The big difference is the `array(..., ndmin=)` is unconditional.

The change you see here is just in the second commit, and looks like a clear bug to me. I also suspect that no one ever used this code path anyway, based on that

**Comment by charris** [comment]
OK, the tests pass, so let's give it a shot. Thanks Eric.

**Pull request merged**